base/MAPL_ISO8601_Time.F90

```diff
@@ -0,0 +1,720 @@
+   integer, parameter :: NUM_MONTHS = UB_MONTH - 1
+   integer, parameter :: LB_DAY = 0
+   integer, parameter :: LB_TIME = -1
+   integer, parameter :: LB_HOUR = LB_TIME
+   integer, parameter :: UB_HOUR = 24
+   integer, parameter :: LB_MINUTE = LB_TIME
+   integer, parameter :: UB_MINUTE = 60
+   integer, parameter :: LB_SECOND = LB_TIME
+   integer, parameter :: UB_SECOND = UB_MINUTE
+   integer, parameter :: LB_MILLISECOND = -1
+   integer, parameter :: UB_MILLISECOND = 1000
+
+   integer, parameter :: TIMEZONE_DEFAULT = 0
+
+   integer, parameter :: ID_WIDTH = 100
+   integer, parameter :: ID_DAY = 1
+   integer, parameter :: ID_MONTH = ID_WIDTH * ID_DAY
+   integer, parameter :: ID_YEAR = ID_WIDTH * ID_MONTH
+
+   type :: date_fields
+      integer :: year
+      integer :: month
+      integer :: day
+      logical :: valid
+   end type date_fields
+
+   interface date_fields
+      module procedure :: construct_date_fields
+   end interface date_fields
+
+   type :: time_fields
+      integer :: hour
+      integer :: minute
+      integer :: second
+      integer :: millisecond
+      integer :: timezone
+      logical :: valid
+   end type time_fields
+
+   interface time_fields
+      module procedure :: construct_time_fields
+   end interface time_fields
+
+   type :: ISO8601Date
+      type(date_fields) :: fields
+   end type ISO8601Date
+
+   interface ISO8601Date
+      module procedure :: construct_ISO8601Date
+   end interface ISO8601Date
+
+   type :: ISO8601Time
+      type(time_fields) :: fields
+      integer :: timezone
+   end type ISO8601Time
+
+   interface ISO8601Time
+      module procedure :: construct_ISO8601Time
+   end interface ISO8601Time
+
+   type :: ISO8601DateTime
+      type(ISO8601Date) :: date
+      type(ISO8601Time) :: time
+   contains
+      procedure, pass(self) :: year
+      procedure, pass(self) :: month
+      procedure, pass(self) :: day
+      procedure, pass(self) :: hour
+      procedure, pass(self) :: minute
+      procedure, pass(self) :: second
+      procedure, pass(self) :: millisecond
+      procedure, pass(self) :: timezone
+   end type ISO8601DateTime
+
+   interface ISO8601DateTime
+      module procedure :: construct_ISO8601DateTime
+   end interface ISO8601DateTime
+
+   type :: ISO8601Duration
+      type(date_fields) :: date
+      type(time_fields) :: time
+   end type ISO8601Duration
+
+   type :: ISO8601TimeInterval
+      type(ISO8601DateTime) :: start_datetime
+      type(ISO8601DateTime) :: end_datetime
+      integer :: repetitions
+   end type ISO8601TimeInterval
+
+contains
+
+
+! SIMPLE CONSTRUCTORS
+
+   pure function construct_date_fields(fields) result(df)
+      integer, dimension(DIM_DF), intent(in) :: fields
+      type(date_fields) :: df
+      df % year = fields(1)
+      df % month = fields(2)
+      df % day = fields(3)
+      df % valid = is_valid_date(fields)
+   end function construct_date_fields
+
+   pure function construct_time_fields(fields) result(tf)
```

Comment:
I will be interested in your reasoning for the signature of this constructor and the preceding one.  Usually one would simply have named scalars rather than an array.  But maybe the use case makes this preferred?  Makes me squeamish about sensitivity to the order of the array elements.